### PR TITLE
Permit Parsing Of Leading Zeros

### PR DESCRIPTION
**Pull request by isomarcte**
This commit changes `BiggerDecimal.parseBiggerDecimalUnsafe` to permit parsing of leading '0' characters, and by extension allows all of our numeric codecs to tolerate decoding `JsonString` values which represent numbers with leading zeros. This is more in line with the `fromString` methods on the various numeric classes.

Note, this change should be given some serious scrutiny. It appears that the original authors had at some time explicitly forbade this behavior, as noted by the existence of the `AFTER_ZERO` state as well as the explicit test case for invalid input including the `String` `"01"`.

I've searched around the issues and I couldn't find any reasoning for this decision, so I still think being lenient here is the right choice, but please double check me.

### CODE REVIEW

**Comment by isomarcte** [comment]
@zarthross I can't seem to reproduce that failure locally, and I don't think it is related to this PR.

**Comment by armanbilge** [comment]
Perhaps it was https://github.com/circe/circe/issues/1911 you saw?

**Comment by isomarcte** [comment]
@zarthross thoughts here?

File: modules/numbers/shared/src/main/scala/io/circe/numbers/BiggerDecimal.scala [hunk 1]
```diff
@@ -263,132 +262,115 @@ object BiggerDecimal {
...
-        if (i >= len) FAILED
-        else {
-          if (input.charAt(i) != '0') START
-          else {
-            i = i + 1
-            AFTER_ZERO
+      var parsedNonLeadingDigit: Boolean = false
+
+      if (i >= len) {
+        null
```
Comment (zarthross) [inline modules/numbers/shared/src/main/scala/io/circe/numbers/BiggerDecimal.scala:268]: I don't mind the short cut here, but we need to make it more clear this is a FAILED state.
```suggestion
        null // state = FAILED
```

File: modules/numbers/shared/src/main/scala/io/circe/numbers/BiggerDecimal.scala [hunk 2]
```diff
@@ -4,6 +4,7 @@ import java.io.Serializable
 import java.lang.StringBuilder
 import java.math.{ BigDecimal, BigInteger }
 import scala.annotation.switch
+import scala.annotation.tailrec
```
Comment (zarthross) [inline modules/numbers/shared/src/main/scala/io/circe/numbers/BiggerDecimal.scala:7]: This isn't needed:

```suggestion
```

**Review by zarthross — APPROVED** [review]
Looks good, just 2 minor things I'd like fixed.

File: modules/numbers/shared/src/main/scala/io/circe/numbers/BiggerDecimal.scala (hunk 1 above)
Comment (isomarcte) [reply modules/numbers/shared/src/main/scala/io/circe/numbers/BiggerDecimal.scala:268]: Comment added, FYI that is also listed in the ScalaDoc. 

I don't love using `null` here. I don't think the single boxing of one `Either` is a problematic cost, but this is most similar to the existing API.

File: modules/numbers/shared/src/main/scala/io/circe/numbers/BiggerDecimal.scala (hunk 2 above)
Comment (isomarcte) [reply modules/numbers/shared/src/main/scala/io/circe/numbers/BiggerDecimal.scala]: :+1:

**Review by zmccoy — APPROVED** [review]
Other than the minor naming nitpick I think this looks good.

File: modules/numbers/shared/src/main/scala/io/circe/numbers/BiggerDecimal.scala [hunk 3]
```diff
@@ -263,132 +261,115 @@ object BiggerDecimal {
...
       var i = if (input.charAt(0) == '-') 1 else 0
-
-      var state =
-        if (i >= len) FAILED
-        else {
-          if (input.charAt(i) != '0') START
-          else {
-            i = i + 1
-            AFTER_ZERO
+      var parsedNonLeadingDigit: Boolean = false
```
[the remark this answers is not in the document]
Comment (isomarcte) [reply modules/numbers/shared/src/main/scala/io/circe/numbers/BiggerDecimal.scala]: updated

**Comment by travisbrown** [comment]
For the record I think accepting non-standard JSON is a mistake (except maybe in some rare cases via a custom parser, but definitely not the default parser).

**Comment by zarthross** [comment]
Hey Travis, while I would generally agree with you, this is only impacting the string based json numbers decoding (which I don’t think is in the spec), and not json parsing.  I think this change is fine, especially given the stdlib parse methods like `BigDecimal.apply` accept the leading zeros just fine when handling strings.

**Pull request merged**